srtcore/core.cpp

```diff
@@ -7856,7 +7825,32 @@ int srt::CUDT::sendCtrlAck(CPacket& ctrlpkt, int size)
     // There are new received packets to acknowledge, update related information.
     if (CSeqNo::seqcmp(ack, m_iRcvLastAck) > 0)
     {
-        ackDataUpTo(ack);
+        // Sanity check if the "selected ACK" points to a sequence
+        // in the past for the buffer. This SHOULD NEVER HAPPEN because
+        // on drop the loss records should have been removed, and the last received
+        // sequence also can't be in the past towards the buffer.
+
+        // NOTE: This problem has been observed when the packet sequence
+        // was incorrectly removed from the receiver loss list. This should
+        // then stay here as a condition in order to detect this problem,
+        // should it happen in the future.
+        if (CSeqNo::seqcmp(ack, m_pRcvBuffer->getStartSeqNo()) < 0)
+        {
+            // Log and DO NOT UPDATE. Count on that next time
+            // the found out ack will be better.
+
+            LOGC(xtlog.Error,
+                    log << CONID() << "ack: IPE: invalid ACK from %" << m_iRcvLastAck << " to %" << ack << " ("
+                    << CSeqNo::seqoff(m_iRcvLastAck, ack) << " packets) buffer=%" << m_pRcvBuffer->getStartSeqNo());
```

Comment:
Trying to find a way to make the log message more clear.
```suggestion
            LOGC(xtlog.Error,
                    log << CONID() << "sendCtrlAck: IPE: invalid ACK position %" << ack
                    ", last ACK %" << m_iRcvLastAck << " ("
                    << CSeqNo::seqoff(m_iRcvLastAck, ack) << " packets) buffer=%" << m_pRcvBuffer->getStartSeqNo());
```

Reply:
I don't understand what is unclear here. ACK position is attempted to be shifted from %(m_iRcvLastAck) to %(ack). This actually isn't anything new, it's the expanded the deleted `ackDataUpTo` function. I expanded it because it was extremely short and did nothing beside a single assignment and some logging, while accessing the buffer that should be under lock better be done directly inside the critical seciton.
```
void srt::CUDT::ackDataUpTo(int32_t ack)
{
    const int acksize SRT_ATR_UNUSED = CSeqNo::seqoff(m_iRcvLastSkipAck, ack);
    if (acksize < 0)
    {
        LOGC(xtlog.Error,
             log << CONID() << "ackDataUpTo: IPE: invalid ACK from %" << m_iRcvLastSkipAck << " to %" << ack << " ("
                 << acksize << " packets)");
        return;
    }
    HLOGC(xtlog.Debug,
          log << CONID() << "ackDataUpTo: %" << m_iRcvLastSkipAck << " -> %" << ack << " (" << acksize << " packets)");

    m_iRcvLastAck = ack;
    m_iRcvLastSkipAck = ack;
}
```

Reply:
Ah, right. `sendCtrlAck` should definitely be mentioned here.